Document first-time developer setup, add conda option

zmbc opened the pull request:
### Type of PR

- [ ] BUG
- [ ] FEAT
- [ ] MAINT
- [x] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?

https://github.com/moj-analytical-services/splink/discussions/2019

### Give a brief description for the solution you have provided

More or less followed the outline I gave in https://github.com/moj-analytical-services/splink/discussions/2019#discussioncomment-8794530, except I reordered some things.

### PR Checklist

- [ ] Added documentation for changes
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [ ] Added tests (if appropriate)
- [ ] Updated CHANGELOG.md (if appropriate)
- [ ] Made changes based off the latest version of Splink
- [ ] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

zmbc commented on CONTRIBUTING.md:
```diff
@@ -45,28 +44,35 @@ There are a number of ways to get involved:
-* [Adding tests](./docs/dev_guides/changing_splink/testing.md) to ensure your code works as expected. These will be run through GitHub Actions when a PR is opened.
-* [Linting](./docs/dev_guides/changing_splink/lint_and_format.md) to ensure that code is styled consistently.
+
+- [Adding tests](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/testing.html) to ensure your code works as expected. These will be run through GitHub Actions when a PR is opened.
```
These links were broken in the hosted version, though they were working in the GitHub version. I didn't see a way to make them relative and still work in both places.

ThomasHepworth replied:
I think it's slightly nicer to link directly to the docs anyway. 

On this though (and something I'll make an issue on), we should check that the links in both our README and Contributing guide are valid. We have a [check](https://github.com/moj-analytical-services/splink/blob/master/.github/workflows/documentation.yml#L116C1-L120C56) in place for HTML files, which we should be able to extend to these markdown files.

ThomasHepworth replied:
Ah, I think we are checking them, we just have a hacky solution in place.

That hacky solution is also why you were probably encountering issues - https://github.com/ADBond/splink/blob/179a3f67de3c1eaab1bea97214082146f46c6547/scripts/preprocess_markdown_includes.py#L48/

zmbc replied:
Ah okay! I think with this change, I can remove the hack. Let me check this.

zmbc replied:
I don't see the file you linked in the `master` branch 🤔

RobinL commented on docs/dev_guides/index.md:
```diff
@@ -9,11 +9,10 @@ We encourage contributions from all users. Whether it be adding a feature, fixin
-Thank you for your interest in contributing to Splink! If this is your first time working with Splink, check our [Contributors Guide](/CONTRIBUTING.html).
+Thank you for your interest in contributing to Splink! If this is your first time working with Splink, check our [Contributors Guide](./CONTRIBUTING.md).
 
 When making changes to Splink, there are a number of common operations that developers need to perform. The guides below lay out some of these common operations, and provides scripts to automate these processes. These include:
```
Should the first point in these bullets link to the new development quickstart page?

RobinL commented on docs/dev_guides/index.md:
```diff
@@ -9,11 +9,10 @@ We encourage contributions from all users. Whether it be adding a feature, fixin
 When making changes to Splink, there are a number of common operations that developers need to perform. The guides below lay out some of these common operations, and provides scripts to automate these processes. These include:
 
-* [Building a Virtual Environment](./changing_splink/building_env_locally.md) - to replicate the conditions when Splink is installed by users.
 * [Linting and Formatting](./changing_splink/lint_and_format.md) - to ensure consistent code style and to reformat code, where possible.
```
Should the first bullet in this list point to the new quickstart page?

RobinL commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -30,6 +27,11 @@ or alternatively
+!!! information Note
+
+    If you do not have an AWS account, or have not set up Splink to use it, you will need to add
+    `--ignore tests/test_full_example_athena.py` to every `pytest` command you run.
```
This is supposed to be handled automatically, but perhaps it's not working.

Specifically, [this](https://github.com/moj-analytical-services/splink/blob/bfaf8d142f7aad03f281b6324dd530e202eaec9a/tests/test_full_example_athena.py#L38) should cause the tests to be skipped if the test cannot create a valid connection to Athena

RobinL reviewed (commented):
Thanks very much this is great.  A few minor comments.

I going to ask Tom to look at this too, just to get a second pair of eyes on a couple parts of this, but overall, very happy with how it's looking.  A big improvement!

RobinL commented:
@ThomasHepworth would you mind having a quick look at this?  I've gone through in detail, but I think it'd be good to get a second pair of eyes specifically on [docs/dev_guides/changing_splink/development_quickstart.md](https://github.com/moj-analytical-services/splink/pull/2083/files#diff-dd47ac575c96624a8cc364857ccc27260d8f115232912e721018546532562480)

I have:
- Built the docs locally and they look fine to me
- Worked through the conda instructions, got the conda environment working and tests running within it

So I think it's just a sense check that's needed

ThomasHepworth commented:
Thank you for doing all of this! 

It's really useful for us to get input and feedback from people coming to Splink from outside our bubble. I'll check it over now and leave a few comments.

ThomasHepworth commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -30,6 +27,11 @@ or alternatively
 
+!!! information Note
+
+    If you do not have an AWS account, or have not set up Splink to use it, you will need to add
```
It might be better to tweak the logic on this function. I've added a decorator in our [linkage repo](https://github.com/moj-analytical-services/data_linking/blob/main/tests/communicate_with_s3.py#L16) that detects whether you're logged in and able to access s3 and skips if not and seems to be more stable.

When I get to this step I'll briefly check it and create an issue if it seems to be worth the effort.

zmbc replied:
The linkage repo is not public; I will apply Robin's suggested change for now and see if it works for me.

zmbc replied:
It does 🎉 -- I'll leave further improvements to a future PR.

ThomasHepworth commented on docs/dev_guides/changing_splink/testing.md:
```diff
@@ -101,11 +103,14 @@ pytest -W ignore -q -x -m duckdb tests/test_estimate_prob_two_rr_match.py
 ??? tip "Running tests with docker 🐳"
 
-    If you want to test Splink against a specific version of python, the easiest method is to utilise docker 🐳.
+    Testing Splink against a specific version of Python can be useful
+    if you're using py > 3.9.10 (which is currently in use in our tests github action)
+    and need to run a secondary set of tests.
```
While this isn't technically incorrect (we are still only running Splink against versions of python <= 3.9.10 in our matrix tests), I think we should probably pre-emptively reword this.

Something along the lines of:
```suggestion
    Testing Splink against a specific version of Python, especially newer versions not included in our GitHub Actions, is vital for identifying compatibility issues 
    early and reviewing errors reported by users.
```

ThomasHepworth reviewed (commented):
Thanks again Zeb! This is a great step forward in this department.

See a few minor comments and discussion points below. I've also had a word with Robin espousing the use of conda and I believe he's going to comment on that side of things.

RobinL commented:
Tom and I had a chat about this earlier.  We both agree it's a great improvement.  

There's one aspect both of us picked up on:  Whilst the quickstart page contains options for both conda and non-conda, the way its layed out implicitly suggests that conda is the recommended option.  This didn't feel quite right because none of the core dev team use conda.  

You'll see I've done a [draft PR](https://github.com/zmbc/splink/pull/1) to move things around a little and include a (hopefully not too controversial) line on whether re recommend conda or not.

zmbc commented on docs/dev_guides/changing_splink/development_quickstart.md:
```diff
@@ -0,0 +1,213 @@
+
+If you **did** use conda, then each time you open a terminal to develop
+Splink, after navigating to the repository directory, run `mamba activate splink`
+and then `poetry shell`.
```
I'll generalize this a bit for other shells (and platforms).

zmbc commented on scripts/conda/development_setup_with_conda.sh:
```diff
@@ -0,0 +1,37 @@
+then
+    curl -L -O "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
+    bash Miniforge3-$(uname)-$(uname -m).sh -b
+    mamba config --set auto_activate_base false
```
I think there is a cleaner way to do this, I will test this more closely.

zmbc replied:
First of all -- apologies for not testing this code in the first place. I thought it would require me to uninstall conda completely, which would be a hassle, but I simply needed to remove it from my PATH. I have addressed this in a different way, the one recommended by the Miniforge docs, and tested that it works now.

zmbc commented on docs/dev_guides/changing_splink/managing_dependencies.md:
Good point. I will propose new titles for both pages.

zmbc replied:
Ok, on reviewing this more, I think I like the title "Managing Dependencies with Poetry" but would recommend "Dependency Compatibility Policy" for the second page you linked.

ThomasHepworth replied:
Sorry Zeb, still catching up with everything else at work.

Both suggestions above sound sensible and should reduce confusion.

zmbc commented on scripts/create_venv.sh:
If we keep it, I think we should at least fix it to [not install poetry within the venv](https://github.com/moj-analytical-services/splink/discussions/2019#discussioncomment-8669927).

It sounds like you all use this script instead of `poetry install` because poetry does not support having multiple venvs at once. I am slightly curious why you need multiple?

ThomasHepworth replied:
Aha, I hadn't seen that thread with comments from Andy and Robin. Given that, I'm happy to remove it. If anyone complains, we point them to the updated guidance.

zmbc commented on scripts/conda/development_environment.yaml:
```diff
@@ -0,0 +1,14 @@
+  - openjdk<18 # Required for Spark; Spark 3.4 only supports older Java
+  - postgresql
+  - pip:
+      - poetry==1.4.2 # TODO: Unpin
```
Yes -- on `master`, the lock file was generated with 1.4.2 and does not work with recent versions of poetry (see #1986). I can follow up with a PR to update this and the lockfile.

zmbc commented on docs/dev_guides/index.md:
```diff
@@ -9,11 +9,10 @@ We encourage contributions from all users. Whether it be adding a feature, fixin
 When making changes to Splink, there are a number of common operations that developers need to perform. The guides below lay out some of these common operations, and provides scripts to automate these processes. These include:
 
-* [Building a Virtual Environment](./changing_splink/building_env_locally.md) - to replicate the conditions when Splink is installed by users.
 * [Linting and Formatting](./changing_splink/lint_and_format.md) - to ensure consistent code style and to reformat code, where possible.
```
I don't see a ton of utility in this list. "Linting and Formatting" and "Testing" are already linked in the contributor's guide section about contributing code (the only type of contribution to which they are relevant). Likewise with building the docs locally.

Releasing a new package version is nice to have documented, but calling it out for new contributors seems unnecessary -- they almost certainly wouldn't have the permissions to do this.

And contributing to the blog seems like perhaps it should just be its own section in the contributor's guide.

RobinL replied:
Yeah agreed, in fact the whole page feels unnecessarily repetitious and could probably be dropped entirely.

We probably need to review it separately.  Just want to make sure the quickstart is as obvious as possible - feel free to do whatever you think's best!

zmbc replied:
Sounds good and happy to help with that. I think in this PR I won't put the quickstart in the list since it is not a "common operation that developers need to perform."

zmbc commented:
Thank you both for your reviews! I believe I've addressed all the comments now. See my comments on the [de-emphasize conda PR](https://github.com/zmbc/splink/pull/1) -- if they look good to you, I can merge those changes into this branch.

RobinL commented:
Thanks @zmbc  yeah happy for you to go ahead and merge in the [de-emphasize conda PR](https://github.com/zmbc/splink/pull/1) and after that we should be good to go.

Thanks again!

zmbc commented:
@ThomasHepworth I think this is just waiting for follow-up on your two unresolved comments above. Once we get those ironed out, it should be good to go!

ThomasHepworth commented on mkdocs.yml:
```diff
@@ -234,11 +234,12 @@ nav:
         - Linting and Formatting: "dev_guides/changing_splink/lint_and_format.md"
         - Testing: "dev_guides/changing_splink/testing.md"
         - Building Docs: "dev_guides/changing_splink/build_docs_locally.md"
+        - Managing Dependencies: "dev_guides/changing_splink/managing_dependencies.md"
```
From - https://github.com/moj-analytical-services/splink/pull/2083

```suggestion
        - Managing Dependencies with Poetry: "dev_guides/changing_splink/managing_dependencies.md"
```

ThomasHepworth commented:
@zmbc after you've tweaked the `mkdocs.yml` file, I'm happy to approve this from my end.

Thanks again for your work on this!

zmbc commented:
Addressed the comments, and I believe my changes should now fix the CI failures, except that the mkdocs CI seems to not like that this PR is from a fork.
